backend/drm.c

```diff
@@ -0,0 +1,300 @@
+                       &CONN(tx)->connector_id, 1, &mode) != 0) {
+        log_error("Failed to set CRTC");
+        goto bail_crtc;
+    }
+
+    return true;
+
+bail_crtc:
+    drmModeFreeCrtc(CRTC(tx));
+bail_mmap:
+    munmap(tx->fb_base, tx->fb_len);
+    drmModeRmFB(tx->drm_dri_fd, tx->fb_id);
+bail_conn:
+    drmModeFreeConnector(CONN(tx));
+bail_res:
+    drmModeFreeResources(RES(tx));
+    return false;
+}
+
+twin_context_t *twin_drm_init(int width, int height)
+{
+    /* Get environment variable to execute */
+    char *drm_dri_path = getenv(DRM_DRI_NAME);
+    if (!drm_dri_path) {
+        log_info("Environment variable $DRI_CARD not set, use %s by default",
+                 DRM_DRI_DEFAULT);
+        drm_dri_path = DRM_DRI_DEFAULT;
+    }
+
+    twin_context_t *ctx = calloc(1, sizeof(twin_context_t));
+    if (!ctx)
+        return NULL;
+
+    ctx->priv = calloc(1, sizeof(twin_drm_t));
+    if (!ctx->priv)
+        return NULL;
+
+    twin_drm_t *tx = ctx->priv;
+
+    /* Open the DRM driver */
+    tx->drm_dri_fd = open(drm_dri_path, O_RDWR);
+    if (tx->drm_dri_fd < 0) {
+        log_error("Failed to open %s", drm_dri_path);
+        goto bail;
+    }
+
+    /* Set up virtual terminal environment */
+    if (!twin_vt_setup(&tx->vt_fd)) {
+        goto bail_dri_card_fd;
+    }
+
+    /* Apply configurations to the DRM driver*/
+    if (!twin_drm_apply_config(tx)) {
+        log_error("Failed to apply configurations to the DRM driver");
+        goto bail_vt_fd;
+    }
+
+    /* Create TWIN screen */
+    ctx->screen =
+        twin_screen_create(width, height, NULL, _twin_drm_put_span, ctx);
+
+    /* Create Linux input system object */
+    tx->input = twin_linux_input_create(ctx->screen);
+    if (!tx->input) {
+        log_error("Failed to create Linux input system object");
+        goto bail_screen;
+    }
+
+    /* Setup file handler and work functions */
+    twin_set_work(twin_drm_work, TWIN_WORK_REDISPLAY, ctx);
+
+    return ctx;
+
+bail_screen:
+    twin_screen_destroy(ctx->screen);
+bail_vt_fd:
+    close(tx->vt_fd);
+bail_dri_card_fd:
+    close(tx->drm_dri_fd);
+bail:
+    free(ctx->priv);
+    free(ctx);
+    return NULL;
+}
+
+static void twin_drm_configure(twin_context_t *ctx)
+{
+    int width, height;
+    twin_drm_t *tx = PRIV(ctx);
+    twin_drm_get_screen_size(tx, &width, &height);
+    twin_screen_resize(SCREEN(ctx), width, height);
+}
+
+static void twin_drm_exit(twin_context_t *ctx)
+{
+    if (!ctx)
+        return;
+
+    twin_drm_t *tx = PRIV(ctx);
+    twin_vt_mode(tx->vt_fd, KD_TEXT);
+    drmModeFreeCrtc(CRTC(tx));
+    munmap(tx->fb_base, tx->fb_len);
+    drmModeRmFB(tx->drm_dri_fd, tx->fb_id);
+    drmModeFreeConnector(CONN(tx));
+    drmModeFreeResources(RES(tx));
+    twin_linux_input_destroy(tx->input);
+    close(tx->vt_fd);
+    close(tx->drm_dri_fd);
+    free(ctx->priv);
+    free(ctx);
+}
+
+/* Register the Linux DRM backend */
+
+const twin_backend_t g_twin_backend = {
```

Comment:
You should add `poll` operation as it is required in recent `main` branch.

Reply:
As the original implementation, we creates another thread to handle the events in function [`twin_linux_evdev_thread`](https://github.com/sysprog21/mado/blob/35ddf50ac4c7fe61b8b3c963f80dabceeb7783fe/backend/linux_input.c#L301). So do you suggest refactoring [linux_input.c#L255~278](https://github.com/sysprog21/mado/blob/35ddf50ac4c7fe61b8b3c963f80dabceeb7783fe/backend/linux_input.c#L255)? 

Or we just assign NULL to `.poll`?
```c
/* Register the Linux DRM backend */

const twin_backend_t g_twin_backend = {
    .init = twin_drm_init,
    .poll = NULL,  /*without implementation*/
    .configure = twin_drm_configure,
    .exit = twin_drm_exit,
};
```

Reply:
@alanjian85, can you comment this?